os/fs/smartfs/smartfs_smart.c

```diff
@@ -1912,54 +1961,83 @@ int smartfs_rename(struct inode *mountpt, const char *oldrelpath, const char *ne
 
 		mode = oldentry.flags & SMARTFS_DIRENT_MODE;
 		type = oldentry.flags & SMARTFS_DIRENT_TYPE;
 
-		ret = smartfs_createentry(fs, newparentdirsector, newfilename, type, mode, &newentry, oldentry.firstsector, NULL);
-		if (ret != OK) {
+		/* Find an available invalid/empty entry to write the new entry */
+		ret = smartfs_find_availableentry(fs, newparentdirsector, &f_newentry, &is_empty);
+		if (ret != OK && ret != -ENOENT) {
+			fdbg("\nfind entry encountered a problem, cannot find entry for writing\n");
 			goto errout_with_semaphore;
 		}
 
-		/* Now mark the old entry as inactive */
-
-		readwrite.logsector = oldentry.dsector;
-		readwrite.offset = 0;
-		readwrite.count = fs->fs_llformat.availbytes;
-		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
-		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
-		if (ret < 0) {
-			fdbg("Error %d reading sector %d data\n", ret, oldentry.dsector);
-			goto errout_with_semaphore;
-		}
-		// bug fix for rename
-		tmp_pntr = (uint8_t *)&fs->fs_rwbuffer[oldentry.doffset];
-		tmp_flag = tmp_pntr[0];
-		tmp_flag |= (tmp_pntr[1] << 8);
+		/* Check if it is a simple rename operation, i.e. in the same directory */
+		if (oldparentdirsector == newparentdirsector) {
+#ifndef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+			if ((ret == OK) && is_empty) {
+				/* The only case where not overwriting the same entry for renaming works better */
+				f_newentry.firstsector = oldentry.firstsector;
+				ret = smartfs_writeentry(fs, f_newentry, newfilename, type, mode, &newentry);
+				if (ret != OK) {
+					fdbg("\nError writing new entry\n");
+					goto errout_with_semaphore;
+				}
 
-#if CONFIG_SMARTFS_ERASEDSTATE == 0xFF
-		tmp_flag &= ~SMARTFS_DIRENT_ACTIVE;
-#else
-		tmp_flag |= SMARTFS_DIRENT_ACTIVE;
+				goto invalidate_entry;
+			} else
 #endif
-		tmp_pntr[0] = (uint8_t)(tmp_flag & 0x00FF);
-		tmp_pntr[1] = (uint8_t)((tmp_flag >> 8) & 0x00FF);
+			{
+				/* No emtpy entry was found available in one of the chained parent sectors */
+				/* We will not use any new entry found, we will overwrite the existing entry */
+				ret = smartfs_writeentry(fs, oldentry, newfilename, type, mode, &newentry);
+				if (ret != OK) {
+					fdbg("\nError writing new entry\n");
+					goto errout_with_semaphore;
+				}
+				/* Old entry doesn't have to be invalidated */
+				goto errout_with_semaphore;
+			}
+		}
 
-		/* Now write the updated flags back to the device */
+		if (ret == -ENOENT) {
+			newparentdirsector = f_newentry.dsector;
+			ret = smartfs_createentry(fs, newparentdirsector, &f_newentry);
+			if (ret != OK) {
+				fdbg("\ncreate entry failed, couldn't allocate new entry\n");
+				goto errout_with_semaphore;
+			}
+		}
 
-		readwrite.offset = oldentry.doffset;
-		readwrite.count = sizeof(uint16_t);
-		readwrite.buffer = (uint8_t *)tmp_pntr;
+		f_newentry.firstsector = oldentry.firstsector;
+		ret = smartfs_writeentry(fs, f_newentry, newfilename, type, mode, &newentry);
```

Comment:
This is corner case.
If power failed after this line, we have 2 entry which points same first sector...
How we realize which latest?
We need discuss it.
Actually my idea is use utc in entry header,
if CONFIG_RTC is enabled, we can use RTC, otherwise sequence number.
For example. we want rename /mnt/a to /mnt/test/a and it's first sector is logical 20.
/mnt/a entry has low entry->utc then /mnt/test/a's entry->utc
During the fs_recover after journal recover, if any logical sector chained to more than 1 time,
Then invalidate old one.

Reply:
Yes, as discussed,
Seq number + fs_recovery will do the job.